Hibernate ORM graphql panache test fails if time zone is west of GMT

### Describe the bug

Test fails if time zone is west of GMT:

```
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3.879 s <<< FAILURE! -- in io.quarkus.it.hibertnate.orm.graphql.panache.HibernateOrmGraphQLPanacheTest
[ERROR] io.quarkus.it.hibertnate.orm.graphql.panache.HibernateOrmGraphQLPanacheTest.testEndpoint -- Time elapsed: 0.947 s <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
JSON path data.authors.dob doesn't match.
Expected: iterable containing ["1821-11-11"]
  Actual: <[1821-11-10]>

	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:73)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:108)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:57)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:512)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
	at io.restassured.internal.ResponseSpecificationImpl.validateResponseIfRequired(ResponseSpecificationImpl.groovy:696)
	at io.restassured.internal.ResponseSpecificationImpl.this$2$validateResponseIfRequired(ResponseSpecificationImpl.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:198)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:62)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at io.restassured.internal.ResponseSpecificationImpl.body(ResponseSpecificationImpl.groovy:270)
	at io.restassured.specification.ResponseSpecification$body$1.callCurrent(Unknown Source)
	at io.restassured.internal.ResponseSpecificationImpl.body(ResponseSpecificationImpl.groovy:117)
	at io.restassured.internal.ValidatableResponseOptionsImpl.body(ValidatableResponseOptionsImpl.java:244)
	at io.quarkus.it.hibertnate.orm.graphql.panache.HibernateOrmGraphQLPanacheTest.testEndpoint(HibernateOrmGraphQLPanacheTest.java:63)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:1013)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:827)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

Setting e.g. `TZ=Europe/Zurich` makes the problem go away.

### Expected behavior

Test should pass in all local time zones.

### Actual behavior

The date is not mapped correctly, presumably because it is taking a `ZonedDateTime` from the local zone and then getting the date after mapping it to GMT.

### How to Reproduce?

Run `TZ=America/Chicago mvn install -rf :quarkus-integration-test-hibernate-orm-graphql-panache` after a clean build e.g. `mvn clean && mvn install -DskipTests`.

### Output of `uname -a` or `ver`

Darwin xxxx 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:45 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6020 arm64

### Output of `java -version`

openjdk 17.0.9 2023-10-17 OpenJDK Runtime Environment Temurin-17.0.9+9 (build 17.0.9+9) OpenJDK 64-Bit Server VM Temurin-17.0.9+9 (build 17.0.9+9, mixed mode)

### Quarkus version or git rev

0d90b7c0b82262ed38a00ac8b34385f19c88dd9a

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546) Maven home: /Users/david/local/apache-maven Java version: 17.0.9, vendor: Eclipse Adoptium, runtime: /Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home Default locale: en_US, platform encoding: UTF-8 OS name: "mac os x", version: "14.1.2", arch: "aarch64", family: "mac"

### Additional information

_No response_

--- a/integration-tests/hibernate-orm-graphql-panache/src/main/java/io/quarkus/it/hibertnate/orm/graphql/panache/Author.java
+++ b/integration-tests/hibernate-orm-graphql-panache/src/main/java/io/quarkus/it/hibertnate/orm/graphql/panache/Author.java
@@ -4,11 +4,14 @@ import java.time.LocalDate;
 
 import jakarta.persistence.Entity;
 
+import org.hibernate.annotations.JdbcType;
+
 import io.quarkus.hibernate.orm.panache.PanacheEntity;
 
 @Entity
 public class Author extends PanacheEntity {
 
     public String name;
+    @JdbcType(LocalDateJdbcType.class)
     public LocalDate dob;
 }

--- a/integration-tests/hibernate-orm-graphql-panache/src/main/java/io/quarkus/it/hibertnate/orm/graphql/panache/LocalDateJdbcType.java
+++ b/integration-tests/hibernate-orm-graphql-panache/src/main/java/io/quarkus/it/hibertnate/orm/graphql/panache/LocalDateJdbcType.java
@@ -1,0 +1,104 @@
+package io.quarkus.it.hibertnate.orm.graphql.panache;
+
+import java.io.Serial;
+import java.sql.CallableStatement;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Types;
+import java.time.LocalDate;
+
+import jakarta.persistence.TemporalType;
+
+import org.hibernate.type.descriptor.ValueBinder;
+import org.hibernate.type.descriptor.ValueExtractor;
+import org.hibernate.type.descriptor.WrapperOptions;
+import org.hibernate.type.descriptor.java.JavaType;
+import org.hibernate.type.descriptor.jdbc.BasicBinder;
+import org.hibernate.type.descriptor.jdbc.BasicExtractor;
+import org.hibernate.type.descriptor.jdbc.JdbcLiteralFormatter;
+import org.hibernate.type.descriptor.jdbc.JdbcType;
+import org.hibernate.type.descriptor.jdbc.internal.JdbcLiteralFormatterTemporal;
+import org.hibernate.type.spi.TypeConfiguration;
+
+/**
+ * A "correct" local date JDBC type.
+ */
+public final class LocalDateJdbcType implements JdbcType {
+    @Serial
+    private static final long serialVersionUID = -3375371178728213643L;
+
+    public static final LocalDateJdbcType INSTANCE = new LocalDateJdbcType();
+
+    private LocalDateJdbcType() {
+    }
+
+    @Override
+    public int getJdbcTypeCode() {
+        return Types.DATE;
+    }
+
+    @Override
+    public String getFriendlyName() {
+        return "DATE";
+    }
+
+    @Override
+    public String toString() {
+        return "LocalDateTypeDescriptor";
+    }
+
+    @Override
+    public <T> JavaType<T> getJdbcRecommendedJavaTypeMapping(
+            Integer length,
+            Integer scale,
+            TypeConfiguration typeConfiguration) {
+        return typeConfiguration.getJavaTypeRegistry().getDescriptor(LocalDate.class);
+    }
+
+    @Override
+    public <T> JdbcLiteralFormatter<T> getJdbcLiteralFormatter(JavaType<T> javaType) {
+        return new JdbcLiteralFormatterTemporal<>(javaType, TemporalType.DATE);
+    }
+
+    @Override
+    public Class<?> getPreferredJavaTypeClass(WrapperOptions options) {
+        return LocalDate.class;
+    }
+
+    @Override
+    public <X> ValueBinder<X> getBinder(final JavaType<X> javaType) {
+        return new BasicBinder<>(javaType, this) {
+            @Override
+            protected void doBind(PreparedStatement st, X value, int index, WrapperOptions options) throws SQLException {
+                st.setObject(index, javaType.unwrap(value, LocalDate.class, options));
+            }
+
+            @Override
+            protected void doBind(CallableStatement st, X value, String name, WrapperOptions options)
+                    throws SQLException {
+                st.setObject(name, javaType.unwrap(value, LocalDate.class, options));
+            }
+        };
+    }
+
+    @Override
+    public <X> ValueExtractor<X> getExtractor(final JavaType<X> javaType) {
+        return new BasicExtractor<>(javaType, this) {
+            @Override
+            protected X doExtract(ResultSet rs, int paramIndex, WrapperOptions options) throws SQLException {
+                return javaType.wrap(rs.getObject(paramIndex, LocalDate.class), options);
+            }
+
+            @Override
+            protected X doExtract(CallableStatement statement, int index, WrapperOptions options) throws SQLException {
+                return javaType.wrap(statement.getObject(index, LocalDate.class), options);
+            }
+
+            @Override
+            protected X doExtract(CallableStatement statement, String name, WrapperOptions options) throws SQLException {
+                return javaType.wrap(statement.getObject(name, LocalDate.class), options);
+            }
+        };
+    }
+}

